sswf-core/src/main/scala/com/bazaarvoice/sswf/model/history/HistoryFactory.scala

```diff
@@ -249,10 +249,9 @@ object HistoryFactory {
             val signalName: String = attributes.getSignalName
 
             val endTime: DateTime = new DateTime(h.getEventTimestamp.getTime)
-
+            currentStep = null // null currentStep after signal so it will get properly set when next step hits ActivityTaskScheduled
```

Comment:
RE: 3, the workflow in Doula has subsequent steps. The CREATE_UNSTEADY step is stopped by a timer waiting on a signal, after the signal the workflow executes several more steps so in fact we are testing a workflow with subsequent steps.

RE: 2, no because if currentStep is null the timer is reset at line 94.

RE: 1, I am not sure. I do know that the step following the signaled step (CREATE_STEADY) goes to the ActivityTaskScheduled case.

Reply:
RE: 1, is there some way we can assert/prove using the type system or at least some manual assertion (not Java/scala "assert" that may be removed by the compiler) that the only possible workflow state transition from the signal step goes to the scheduled case? I think that's very important.

RE: 2, yes, assuming point 1 always remains true.

RE: 3, that's a start -- do any of the subsequent steps read / branch on the currentStepStart timer? If not, we should do some kind of test with a state transition to a step that does use currentStepStart and then verify the step behaved as expected. This is really just me asking to be very thorough because the potential impact of invalid state transitions.

Reply:
RE: 1 OK I'll map out the flow.

Reply:
OK I am going to enter this lengthy comment about the state machine (bottom line is the current fix is correct) here and in the ticket:

1) As part of addressing Jona’s CR comments to last night’s PR I was trying to verify that the following sequence of events is what happens for a signaled workflow:

	a) a step that is waiting on a signal returns a WAIT

	b) the WAIT step gets picked up by the TaskActivityCompleted case entry in the HistoryFactory class and inProgress is set to true so the currentStep state variable and currentStepStart state variable are not changed

	c) after the workflow is signaled there is no entry in the history that handles the “signal received”

	c) therefore after the workflow is signaled, it effectively jumps straight to the TaskActivityScheduled step for the next step in the workflow

2) I instrumented every case in the HistoryFactory.from() method and this is the history that you get when you start a Doula workflow and it blocks at the end of CREATE\_UNSTEADY for a signal, then receives the signal and continues the rest of the stream boot workflow:
{code}
=======> case branch == WorkflowExecutionStarted, history entry == {EventTimestamp: Tue Feb 27 01:07:52 CST 2018,EventType: WorkflowExecutionStarted,EventId: 1,WorkflowExecutionStartedEventAttributes: {Input: a6a1c4a67296f4c64a1c4bad4b6bbae0,ExecutionStartToCloseTimeout: 2592000,TaskStartToCloseTimeout: 600,ChildPolicy: TERMINATE,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},WorkflowType: {Name: doula-main,Version: 3.17-SNAPSHOT_local_local_local_mcard-r.local},ParentInitiatedEventId: 0,},}
=======> case branch ==_, history entry == {EventTimestamp: Tue Feb 27 01:07:52 CST 2018,EventType: DecisionTaskScheduled,EventId: 2,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:07:52 CST 2018,EventType: DecisionTaskStarted,EventId: 3,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 2},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:07:53 CST 2018,EventType: DecisionTaskCompleted,EventId: 4,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 2,StartedEventId: 3},}
=======> case branch == ActivityTaskScheduled, history entry == {EventTimestamp: Tue Feb 27 01:07:53 CST 2018,EventType: ActivityTaskScheduled,EventId: 5,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: CREATE_UNSTEADY,Version: 600.600},ActivityId: CREATE_UNSTEADY,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 900,StartToCloseTimeout: 600,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 4,HeartbeatTimeout: 600},}
=======> case branch == ActivityTaskStarted, history entry == {EventTimestamp: Tue Feb 27 01:07:53 CST 2018,EventType: ActivityTaskStarted,EventId: 6,ActivityTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 5},}
=======> case branch == ActivityTaskCompleted, history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: ActivityTaskCompleted,EventId: 7,ActivityTaskCompletedEventAttributes: {Result: {"result":"WAIT","waitSeconds":300,"signals":["local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed"]},ScheduledEventId: 5,StartedEventId: 6},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: DecisionTaskScheduled,EventId: 8,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: DecisionTaskStarted,EventId: 9,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 8},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: DecisionTaskCompleted,EventId: 10,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 8,StartedEventId: 9},}
=======> case branch == TimerStarted (signals), history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: TimerStarted,EventId: 11,TimerStartedEventAttributes: {TimerId: c1e94f3f-cf61-4c3f-b7f4-597edbab3102,Control: WaitForSignals:local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed,StartToFireTimeout: 300,DecisionTaskCompletedEventId: 10},}
=======> case branch == WorkflowExecutionSignaled, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: WorkflowExecutionSignaled,EventId: 12,WorkflowExecutionSignaledEventAttributes: {SignalName: local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed,ExternalInitiatedEventId: 0},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: DecisionTaskScheduled,EventId: 13,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: DecisionTaskStarted,EventId: 14,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 13},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:50 CST 2018,EventType: DecisionTaskCompleted,EventId: 15,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 13,StartedEventId: 14},}
=======> case branch == ActivityTaskScheduled, history entry == {EventTimestamp: Tue Feb 27 01:08:50 CST 2018,EventType: ActivityTaskScheduled,EventId: 16,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: CREATE_STEADY,Version: 600.600},ActivityId: CREATE_STEADY,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 900,StartToCloseTimeout: 600,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 15,HeartbeatTimeout: 600},}
=======> case branch == ActivityTaskStarted, history entry == {EventTimestamp: Tue Feb 27 01:08:50 CST 2018,EventType: ActivityTaskStarted,EventId: 17,ActivityTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 16},}
=======> case branch == ActivityTaskCompleted, history entry == {EventTimestamp: Tue Feb 27 01:08:51 CST 2018,EventType: ActivityTaskCompleted,EventId: 18,ActivityTaskCompletedEventAttributes: {Result: {"result":"SUCCESS","message":"Cohort[false, None, true, None, None]: created 1 table(s) (0 remain): [Table(doula:test:jobs:e2e:31d14580d72461f6,Set(test),None,false,true,None,None)] to [Table(doula:test:jobs:e2e:31d14580d72461f6,Set(test),None,false,true,None,None)]"},ScheduledEventId: 16,StartedEventId: 17},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:51 CST 2018,EventType: DecisionTaskScheduled,EventId: 19,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:51 CST 2018,EventType: DecisionTaskStarted,EventId: 20,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 19},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:52 CST 2018,EventType: DecisionTaskCompleted,EventId: 21,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 19,StartedEventId: 20},}
=======> case branch == ActivityTaskScheduled, history entry == {EventTimestamp: Tue Feb 27 01:08:52 CST 2018,EventType: ActivityTaskScheduled,EventId: 22,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: ISSUE_STREAM_BOOT,Version: 86400.600},ActivityId: ISSUE_STREAM_BOOT,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 86700,StartToCloseTimeout: 86400,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 21,HeartbeatTimeout: 600},}
=======> case branch == ActivityTaskStarted, history entry == {EventTimestamp: Tue Feb 27 01:08:52 CST 2018,EventType: ActivityTaskStarted,EventId: 23,ActivityTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 22},}
=======> case branch == ActivityTaskCompleted, history entry == {EventTimestamp: Tue Feb 27 01:08:52 CST 2018,EventType: ActivityTaskCompleted,EventId: 24,ActivityTaskCompletedEventAttributes: {Result: {"result":"SUCCESS","message":"successful[1] failed[0] notInSource[0] notInPolloi[0] notSteady[0]"},ScheduledEventId: 22,StartedEventId: 23},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:52 CST 2018,EventType: DecisionTaskScheduled,EventId: 25,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:52 CST 2018,EventType: DecisionTaskStarted,EventId: 26,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 25},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:54 CST 2018,EventType: DecisionTaskCompleted,EventId: 27,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 25,StartedEventId: 26},}
=======> case branch == ActivityTaskScheduled, history entry == {EventTimestamp: Tue Feb 27 01:08:54 CST 2018,EventType: ActivityTaskScheduled,EventId: 28,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: WAIT_BOOT,Version: 200.600},ActivityId: WAIT_BOOT,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 500,StartToCloseTimeout: 200,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 27,HeartbeatTimeout: 600},}
=======> case branch == ActivityTaskStarted, history entry == {EventTimestamp: Tue Feb 27 01:08:54 CST 2018,EventType: ActivityTaskStarted,EventId: 29,ActivityTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 28},}
=======> case branch == ActivityTaskCompleted, history entry == {EventTimestamp: Tue Feb 27 01:08:54 CST 2018,EventType: ActivityTaskCompleted,EventId: 30,ActivityTaskCompletedEventAttributes: {Result: {"result":"IN_PROGRESS","message":"waiting on stream boot: 8 estimated splits"},ScheduledEventId: 28,StartedEventId: 29},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:54 CST 2018,EventType: DecisionTaskScheduled,EventId: 31,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:54 CST 2018,EventType: DecisionTaskStarted,EventId: 32,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 31},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:08:56 CST 2018,EventType: DecisionTaskCompleted,EventId: 33,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 31,StartedEventId: 32},}
=======> case branch == TimerStarted (retry), history entry == {EventTimestamp: Tue Feb 27 01:08:56 CST 2018,EventType: TimerStarted,EventId: 34,TimerStartedEventAttributes: {TimerId: c425ca4b-f4b4-437a-9036-53a905d1b266,Control: {"stepInput":{},"stepName":"WAIT_BOOT"},StartToFireTimeout: 20,DecisionTaskCompletedEventId: 33},}
=======> case branch == TimerFired (retry), history entry == {EventTimestamp: Tue Feb 27 01:09:16 CST 2018,EventType: TimerFired,EventId: 35,TimerFiredEventAttributes: {TimerId: c425ca4b-f4b4-437a-9036-53a905d1b266,StartedEventId: 34},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:09:16 CST 2018,EventType: DecisionTaskScheduled,EventId: 36,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:09:16 CST 2018,EventType: DecisionTaskStarted,EventId: 37,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 36},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:09:18 CST 2018,EventType: DecisionTaskCompleted,EventId: 38,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 36,StartedEventId: 37},}
=======> case branch == ActivityTaskScheduled, history entry == {EventTimestamp: Tue Feb 27 01:09:18 CST 2018,EventType: ActivityTaskScheduled,EventId: 39,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: WAIT_BOOT,Version: 200.600},ActivityId: WAIT_BOOT,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 500,StartToCloseTimeout: 200,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 38,HeartbeatTimeout: 600},}
=======> case branch == ActivityTaskStarted, history entry == {EventTimestamp: Tue Feb 27 01:09:18 CST 2018,EventType: ActivityTaskStarted,EventId: 40,ActivityTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 39},}
=======> case branch == ActivityTaskCompleted, history entry == {EventTimestamp: Tue Feb 27 01:09:18 CST 2018,EventType: ActivityTaskCompleted,EventId: 41,ActivityTaskCompletedEventAttributes: {Result: {"result":"SUCCESS","message":"Boot done: bulk 0[0] 1[0] 2[0] 3[0]; stream[0]; trigger[0]"},ScheduledEventId: 39,StartedEventId: 40},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:09:18 CST 2018,EventType: DecisionTaskScheduled,EventId: 42,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == _, history entry == {EventTimestamp: Tue Feb 27 01:09:18 CST 2018,EventType: DecisionTaskStarted,EventId: 43,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 42},}
{code}

3) Notice here that on line 5 of the above log the CREATE\_UNSTEADY step in the Doula workflow is scheduled for execution:
{code}
=======> case branch == ActivityTaskScheduled history entry == {EventTimestamp: Tue Feb 27 01:07:53 CST 2018,EventType: ActivityTaskScheduled,EventId: 5,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: CREATE_UNSTEADY,Version: 600.600},ActivityId: CREATE_UNSTEADY,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 900,StartToCloseTimeout: 600,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 4,HeartbeatTimeout: 600},}
{code}

4) Notice also that the CREATE\_UNSTEADY step is **completed** when it blocks on waiting for a signal in line 7:
{code}
=======> case branch == ActivityTaskCompleted, history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: ActivityTaskCompleted,EventId: 7,ActivityTaskCompletedEventAttributes: {Result: {"result":"WAIT","waitSeconds":300,"signals":["local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed"]},ScheduledEventId: 5,StartedEventId: 6},}
{code}

5) Notice that the signal timeout timer is started in line 11:
{code}
=======> case branch == TimerStarted (signals), history entry == {EventTimestamp: Tue Feb 27 01:07:54 CST 2018,EventType: TimerStarted,EventId: 11,TimerStartedEventAttributes: {TimerId: c1e94f3f-cf61-4c3f-b7f4-597edbab3102,Control: WaitForSignals:local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed,StartToFireTimeout: 300,DecisionTaskCompletedEventId: 10},}
{code}

6) Notice that the workflow is signaled in line 12:
{code}
=======> case branch == WorkflowExecutionSignaled, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: WorkflowExecutionSignaled,EventId: 12,WorkflowExecutionSignaledEventAttributes: {SignalName: local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed,ExternalInitiatedEventId: 0},}
{code}

7) Notice that the next step in the Doula workflow (CREATE\_STEADY) begins at line 16 and there have been no calls to ActivityTaskCompleted or anything else that would reset the currentStep state variable (the first line below is line 12):
{code}
=======> case branch == WorkflowExecutionSignaled, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: WorkflowExecutionSignaled,EventId: 12,WorkflowExecutionSignaledEventAttributes: {SignalName: local-local-default-allison|22lBDCIGTjJFfecatoXjNW1js8MAJnUmsNHAYdpjGGqRU=|25496191952d8ed,ExternalInitiatedEventId: 0},}
=======> case branch == \_, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: DecisionTaskScheduled,EventId: 13,DecisionTaskScheduledEventAttributes: {TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},StartToCloseTimeout: 600},}
=======> case branch == \_, history entry == {EventTimestamp: Tue Feb 27 01:08:49 CST 2018,EventType: DecisionTaskStarted,EventId: 14,DecisionTaskStartedEventAttributes: {Identity: 34260@mcard-r.local,ScheduledEventId: 13},}
=======> case branch == \_, history entry == {EventTimestamp: Tue Feb 27 01:08:50 CST 2018,EventType: DecisionTaskCompleted,EventId: 15,DecisionTaskCompletedEventAttributes: {ScheduledEventId: 13,StartedEventId: 14},}
=======> case branch == ActivityTaskScheduled, history entry == {EventTimestamp: Tue Feb 27 01:08:50 CST 2018,EventType: ActivityTaskScheduled,EventId: 16,ActivityTaskScheduledEventAttributes: {ActivityType: {Name: CREATE_STEADY,Version: 600.600},ActivityId: CREATE_STEADY,Input: {"stepInput":{},"wfInput":"a6a1c4a67296f4c64a1c4bad4b6bbae0"},ScheduleToStartTimeout: 300,ScheduleToCloseTimeout: 900,StartToCloseTimeout: 600,TaskList: {Name: 3.17-SNAPSHOT_local_local_local_mcard-r.local},DecisionTaskCompletedEventId: 15,HeartbeatTimeout: 600},}
{code}

8) What this shows is that the history does not include a “second” ActivityTaskCompleted entry for when the signal is received by the workflow. The WorkflowExecutionSignaled case entry serves as the “completion” of a caught signal and it is here that the currentStep state variable must be reset to ensure that the workflow may proceed normally.